File: test-suite/golden-tests/interop-3/good-pool.R
#!/usr/bin/env Rscript

source("/home/z/.morloc/src/rbase/R/core.R")
source("/home/z/.morloc/src/rbase/data.R")

.morloc_run <- function(f, args){
  fails <- ""
  isOK <- TRUE
  warns <- list()
  notes <- capture.output(
    {
      value <- withCallingHandlers(
        tryCatch(
          do.call(f, args),
          error = function(e) {
            fails <<- e$message;
            isOK <<- FALSE
          }
        ),
        warning = function(w){
          warns <<- append(warns, w$message)
          invokeRestart("muffleWarning")
        }
      )
    },
    type="message"
  )
  list(
    value = value,
    isOK  = isOK,
    fails = fails,
    warns = warns,
    notes = notes
  )
}

# dies on error, ignores warnings and messages
.morloc_try <- function(f, args, .log=stderr(), .pool="_", .name="_"){
  x <- .morloc_run(f=f, args=args)
  location <- sprintf("%s::%s", .pool, .name)
  if(! x$isOK){
    cat("** R errors in ", location, "\n", file=stderr())
    cat(x$fails, "\n", file=stderr())
    stop(1)
  }
  if(! is.null(.log)){
    lines = c()
    if(length(x$warns) > 0){
      cat("** R warnings in ", location, "\n", file=stderr())
      cat(paste(unlist(x$warns), sep="\n"), file=stderr())
    }
    if(length(x$notes) > 0){
      cat("** R messages in ", location, "\n", file=stderr())
      cat(paste(unlist(x$notes), sep="\n"), file=stderr())
    }
  }
  x$value
}

.morloc_unpack <- function(unpacker, x, .pool, .name){
  x <- .morloc_try(f=unpacker, args=list(as.character(x)), .pool=.pool, .name=.name)
  return(x)
}

.morloc_foreign_call <- function(cmd, args, .pool, .name){
  .morloc_try(f=system2, args=list(cmd, args=args, stdout=TRUE), .pool=.pool, .name=.name)
}

# NOTE: bad inlining
# m3 <- function(x0)
# {
#     a1 <- .morloc_foreign_call("python3", list("pool.py", "18", x0), "_", "_")
#     a2 <- rmorlocinternals::mlc_deserialize(a1, '{"list":["numeric"]}');
#     a0 <- .morloc_foreign_call("python3", list("pool.py", "286", x0), "_", "_")
#     a3 <- mlc_fold(a0, 0.0, a2)
#     a4 <- rmorlocinternals::mlc_serialize(a3, '"numeric"')
#     a4
# }
m666 <- function(x,y){
    x1 <- rmorlocinternals::mlc_serialize(x, '"numeric"')
    y1 <- rmorlocinternals::mlc_serialize(y, '"numeric"')
    a0 <- .morloc_foreign_call("python3", list("pool.py", "286", x1, y1), "_", "_")
    a1 <- rmorlocinternals::mlc_deserialize(a0, '"numeric"')
    a1
}
m3 <- function(x0)
{
    a1 <- .morloc_foreign_call("python3", list("pool.py", "18", x0), "_", "_")
    a2 <- rmorlocinternals::mlc_deserialize(a1, '{"list":["numeric"]}');
    a3 <- mlc_fold(m666, 0.0, a2)
    a3
}

args <- as.list(commandArgs(trailingOnly=TRUE))
if(length(args) == 0){
  stop("Expected 1 or more arguments")
} else {
  cmdID <- args[[1]]
  f_str <- paste0("m", cmdID)
  if(exists(f_str)){
    f <- eval(parse(text=paste0("m", cmdID)))
    result <- do.call(f, args[-1])
    cat(result, "\n")
  } else {
    cat("Could not find manifold '", cmdID, "'\n", file=stderr())
  }
}

File: test-suite/golden-tests/interop-3/good-pool.py
#!/usr/bin/env python

import sys
import subprocess
import json
from pymorlocinternals import (mlc_serialize, mlc_deserialize)
from collections import OrderedDict

sys.path = ["/home/z/.morloc/src"] + sys.path

from pybase.core import *
from pybase.data import *

def _morloc_foreign_call(args):
    try:
        sysObj = subprocess.run(
            args,
            stdout=subprocess.PIPE,
            check=True
        )
    except subprocess.CalledProcessError as e:
        sys.exit(str(e))

    return(sysObj.stdout.decode("ascii"))

# NOTE: m286 has native args, it didn't deserialize them, why?
#  def m286(x0, x1):
#      a0 = mlc_add(x0, x1)
#      a1 = mlc_serialize(a0, ("float",None))
#      return(a1)
def m286(x0, x1):
    a0 = mlc_deserialize(x0, ("float",None))
    a1 = mlc_deserialize(x1, ("float",None))
    a2 = mlc_add(a0, a1)
    a3 = mlc_serialize(a2, ("float",None))
    return(a3)

def m18(x0):
    a1 = mlc_deserialize(x0, ("list",(("float",None))));
    a2 = mlc_map(m16, a1)
    a3 = mlc_serialize(a2, ("list",(("float",None))))
    return(a3)

# NOTE: extra argument
#  def m16(x0, x1):
#      a0 = mlc_add(1.0, x1)
#      return(a0)
def m16(x1):
    a0 = mlc_add(1.0, x1)
    return(a0)

if __name__ == '__main__':
    try:
        cmdID = int(sys.argv[1])
    except IndexError:
        sys.exit("Internal error in {}: no manifold id found".format(sys.argv[0]))
    except ValueError:
        sys.exit("Internal error in {}: expected integer manifold id".format(sys.argv[0]))
    try:
        dispatch = {
            286: m286,
            18: m18,
        }
        f = dispatch[cmdID]
    except KeyError:
        sys.exit("Internal error in {}: no manifold found with id={}".format(sys.argv[0], cmdID))

    result = f(*sys.argv[2:])

    print(result)
